formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/Decoder.kt

```diff
@@ -0,0 +1,641 @@
+        }
+        if (elemName == null) {
+            throw CborDecodingException("Expected (tagged) string or number, got nothing for $descriptor")
+        }
+        return elemName to tags
+    }
+
+    private fun decodeElementNameWithTagsLenient(descriptor: SerialDescriptor): Pair<String?, ULongArray?> {
+        var (elemName, cborLabel, tags) = decoder.nextTaggedStringOrNumber()
+        if (elemName == null && cborLabel != null) {
+            elemName = descriptor.getElementNameForCborLabel(cborLabel)
+        }
+        return elemName to tags
+    }
+
+    @OptIn(ExperimentalSerializationApi::class)
+    override fun <T> decodeSerializableValue(deserializer: DeserializationStrategy<T>): T {
+        return if ((decodeByteArrayAsByteString || cbor.configuration.alwaysUseByteString)
+            && deserializer.descriptor == ByteArraySerializer().descriptor
+        ) {
+            @Suppress("UNCHECKED_CAST")
+            decoder.nextByteString(tags) as T
+        } else {
+            decodeByteArrayAsByteString = decodeByteArrayAsByteString || deserializer.descriptor.isInlineByteString()
+            super<AbstractDecoder>.decodeSerializableValue(deserializer)
+        }
+    }
+
+    override fun decodeString() = decoder.nextString(tags)
+
+    override fun decodeNotNullMark(): Boolean = !decoder.isNull()
+
+    override fun decodeDouble() = decoder.nextDouble(tags)
+    override fun decodeFloat() = decoder.nextFloat(tags)
+
+    override fun decodeBoolean() = decoder.nextBoolean(tags)
+
+    override fun decodeByte() = decoder.nextNumber(tags).toByte()
+    override fun decodeShort() = decoder.nextNumber(tags).toShort()
+    override fun decodeChar() = decoder.nextNumber(tags).toInt().toChar()
+    override fun decodeInt() = decoder.nextNumber(tags).toInt()
+    override fun decodeLong() = decoder.nextNumber(tags)
+
+    override fun decodeNull() = decoder.nextNull(tags)
+
+    override fun decodeEnum(enumDescriptor: SerialDescriptor): Int =
+        enumDescriptor.getElementIndexOrThrow(decoder.nextString(tags))
+
+    private fun isDone(): Boolean = !finiteMode && decoder.isEnd() || (finiteMode && readProperties >= size)
+
+    private fun verifyKeyTags(descriptor: SerialDescriptor, index: Int, tags: ULongArray?) {
+        if (cbor.configuration.verifyKeyTags) {
+            descriptor.getKeyTags(index)?.let { keyTags ->
+                if (!(keyTags contentEquals tags)) throw CborDecodingException(
+                    "CBOR tags ${
+                        tags?.joinToString(
+                            prefix = "[",
+                            postfix = "]"
+                        ) { it.toString() }
+                    } do not match expected tags ${
+                        keyTags.joinToString(
+                            prefix = "[",
+                            postfix = "]"
+                        ) { it.toString() }
```

Comment:
Since you copied `it.toString()`from `CborParser.processTags` here, `CborReader` now also requires `@SuppressAnimalSniffer` annotation. But I would prefer if you just add another method to `CborParser` — e.g. `verifyTagsAndThrow(expected: ULongArray, actual: ULongArray)` to reduce the amount of copy-pasted code

Reply:
I hope I understood correctly. See 7ec8e02909e4e79e4f992dbb95ac18cd48b25bcc